sno/sno_repo.py

```diff
@@ -0,0 +1,146 @@
+import struct
+from pathlib import Path
+
+import pygit2
+
+from .repository_version import write_repo_version_config
+from . import repo_files
+
+
+# Utilities for maintaining file structure of a sno repo - not the versioned controlled blobs,
+# but the .git file, .sno directory, .sno/index file. pygit2 mostly takes care of this for us,
+# but since we differ slightly from a standard git repository, there are a few extra things to take care of.
+
+
+def _append_checksum(data):
+    return data + pygit2.hash(data).raw
+
+
+INDEX_VERSION = 2
+EMPTY_GIT_INDEX = struct.pack(">4sII", b"DIRC", INDEX_VERSION, 0)
+
+# Extension name does not start with A-Z => is a required extension.
+LOCKED_SNO_EXTENSION = struct.pack(">4sI", b".sno", 0)
+
+# An empty index file, but exteneded with a required ".sno" extension in the extensions section of the index binary
+# format. (Not the file extension - the filename is simply "index", it has no file extension.)
+# Causes all git commands that would involve the index or working copy to fail with "unsupported extension: .sno" -
+# in that sense it is "locked" to git. Various techniques can be used to unlock it if certain git functionality is
+# needed - eg marking the repository as bare so it is ignored, or removing the unsupported extension.
+
+# See https://git-scm.com/docs/index-format
+LOCKED_EMPTY_GIT_INDEX = _append_checksum(EMPTY_GIT_INDEX + LOCKED_SNO_EXTENSION)
+
+
+def init_repository(repo_root_path, repo_version):
+    """
+    Initialise a new sno repo. A sno repo is basically a git repo, except -
+    - git internals are stored in .sno instead of .git
+      (.git is a file that contains a reference to .sno, this is allowed by git)
+    - datasets are stored in /.sno-dataset/ trees according to a particular dataset format version - see DATASETS_v2.md
+      but, this only matters once we start committing data. When initialising a repo, these are not yet present.
+    - there is a blob called sno.repository.version that contains the dataset format version number - but, this is only
+      written when data is first imported. When initialising a repo, this is not yet present.
+    - there is property in the repo config called sno.repository.version that contains the dataset format version
+      number, which is used until the sno.repository.version blob is written.
+    - the .sno/index file has been extended to stop git messing things up - see LOCKED_EMPTY_GIT_INDEX.
+
+    There can also be properties in the repo config relating to the working copy - named sno.workingcopy.* - but
+    these are not written by this function, it is up to the caller to configure these as needed.
+    See WorkingCopy.write_config()
+    """
+
+    repo = pygit2.init_repository(str(repo_root_path.resolve()))
+    repo = convert_git_repo_to_sno_repo(repo_root_path)
+    write_repo_version_config(repo, repo_version)
+    assert is_sno_repo(repo)
+    return repo
+
+
+def convert_git_repo_to_sno_repo(repo_root_path):
```

Comment:
again, think we should make sure we have a clear recovery off-ramp for this aborting part-way through. 

eg: no index or no reflog config would be annoying and non-obvious, even though `.sno/` is there and it'd otherwise seem to "work"

Maybe just a case of re-ordering the steps?

Reply:
Re-ordered the steps. Moving .git to .sno is now the first step, and creating the .git file that points to .sno is now the last step. Any failure between these two steps now leaves you with a repo that nothing can find unless they specifically know to look in the .sno directory - which neither git or sno do automatically without the .git file.

Of course, we could potentially fail to rename .git to .sno - attempted to ameliorate this by making any command that would call convert_git_repo_to_sno_repo give up before creating a git repo if .git or .sno already exist, which is the most likely reason the rename would fail.